pkg/event/target/store.go

```diff
@@ -18,20 +18,100 @@ package target
 
 import (
 	"errors"
+	"fmt"
+	"strings"
+	"time"
+
 	"github.com/minio/minio/pkg/event"
 )
 
-// ErrLimitExceeded error is sent when the maximum limit is reached.
-var ErrLimitExceeded = errors.New("[Store] The maximum limit reached")
+const retryInterval = 3 * time.Second
+
+// errNotConnected - indicates that the target connection is not active.
+var errNotConnected = errors.New("not connected to target server/service")
+
+// errLimitExceeded error is sent when the maximum limit is reached.
+var errLimitExceeded = errors.New("the maximum store limit reached")
 
-// ErrNoSuchKey error is sent in Get when the key is not found.
-var ErrNoSuchKey = errors.New("[Store] No such key found")
+// errNoSuchKey error is sent in Get when the key is not found.
+var errNoSuchKey = errors.New("no such key found in store")
 
 // Store - To persist the events.
 type Store interface {
 	Put(event event.Event) error
 	Get(key string) (event.Event, error)
-	ListAll() []string
-	Del(key string)
+	ListN(n int) []string
+	Del(key string) error
 	Open() error
 }
+
+// replayEvents - Reads the events from the store and replays.
+func replayEvents(store Store, doneCh <-chan struct{}) <-chan string {
+	var names []string
+	eventKeyCh := make(chan string)
+
+	go func() {
+		retryTimer := time.NewTimer(retryInterval)
+		defer retryTimer.Stop()
+		for {
+			names = store.ListN(100)
+			for _, name := range names {
+				select {
+				case eventKeyCh <- strings.TrimSuffix(name, eventExt):
+					// Get next key.
+				case <-doneCh:
+					close(eventKeyCh)
+					return
+				}
+			}
+
+			if len(names) < 2 {
+				retryTimer.Reset(retryInterval)
+				select {
+				case <-retryTimer.C:
+				case <-doneCh:
+					close(eventKeyCh)
```

Comment:
`close(eventKeyCh)` can be `defer` in this goroutine

Reply:
Done :+1: